src/adapter/src/coord.rs

```diff
@@ -2132,6 +2134,62 @@ impl Coordinator {
             ),
         );
 
+        let builtin_updates_fut = self
+            .bootstrap_builtin_tables(&entries, builtin_table_updates)
+            .await;
+
+        // Destructure Self so we can do some concurrent work.
+        let Self {
+            secrets_controller,
+            catalog,
+            ..
+        } = self;
+
+        // Cleanup orphaned secrets. Errors during list() or delete() do not
+        // need to prevent bootstrap from succeeding; we will retry next
+        // startup.
+        let secrets_cleanup_fut = async move {
+            match secrets_controller.list().await {
+                Ok(controller_secrets) => {
+                    // Fetch all IDs from the catalog to future-proof against other
+                    // things using secrets. Today, SECRET and CONNECTION objects use
+                    // secrets_controller.ensure, but more things could in the future
+                    // that would be easy to miss adding here.
+                    let catalog_ids: BTreeSet<GlobalId> =
+                        catalog.entries().map(|entry| entry.id()).collect();
+                    let controller_secrets: BTreeSet<GlobalId> =
+                        controller_secrets.into_iter().collect();
+                    let orphaned = controller_secrets.difference(&catalog_ids);
+                    for id in orphaned {
+                        info!("coordinator init: deleting orphaned secret {id}");
+                        fail_point!("orphan_secrets");
+                        if let Err(e) = secrets_controller.delete(*id).await {
+                            warn!("Dropping orphaned secret has encountered an error: {}", e);
+                        }
+                    }
+                }
+                Err(e) => warn!("Failed to list secrets during orphan cleanup: {:?}", e),
+            }
+        };
+
+        // Run all of our final steps concurrently.
+        futures::future::join_all([builtin_updates_fut, Box::pin(secrets_cleanup_fut)])
+            .instrument(info_span!("coord::bootstrap::final"))
+            .await;
+
+        info!("coordinator init: bootstrap complete");
+        Ok(())
+    }
+
+    /// Prepares builtin tables for writing by resetting them to a known state
+    /// and appending the given updates.
+    #[allow(clippy::async_yields_async)]
+    #[instrument]
+    async fn bootstrap_builtin_tables(
+        &mut self,
+        entries: &[CatalogEntry],
+        mut builtin_table_updates: Vec<BuiltinTableUpdate>,
+    ) -> BuiltinTableAppendNotify {
         // Advance all tables to the current timestamp
```

Comment:
I _think_ it might not be doing anything, yes, but didn't want to do changes beyond what's necessary. The act of writing to the tables (that is, via txn) also does nothing, fencing wise, or does it? I keep forgetting whether we retry table appends when there's a conflict... 🙈

This whole thing shows once more how read-then-write, with the global (process-wide) write lock is on shaky ground, though.

I'd do one of these:
1. remove the timestamp and appends, keep the current name
2. rename to `bootstrap_tables`, keep the potentially questionable code

wdyt?

Reply:
I think option (2) is the safer approach and we can consider whether it's safe to remove this code in a separate commit.